core/runtime/src/main/java/io/quarkiverse/langchain4j/guardrails/AbstractJsonExtractorOutputGuardrail.java

```diff
@@ -0,0 +1,60 @@
+package io.quarkiverse.langchain4j.guardrails;
+
+import jakarta.inject.Inject;
+
+import org.jboss.logging.Logger;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+
+import dev.langchain4j.data.message.AiMessage;
+
+public abstract class AbstractJsonExtractorOutputGuardrail implements OutputGuardrail {
+
+    @Inject
+    Logger logger;
+
+    @Inject
+    GuardrailsUtils guardrailsUtils;
+
+    protected AbstractJsonExtractorOutputGuardrail() {
+        if (getOutputClass() == null && getOutputType() == null) {
+            throw new IllegalArgumentException("Either getOutputClass() or getOutputType() must be implemented");
+        }
+    }
+
+    @Override
+    public OutputGuardrailResult validate(AiMessage responseFromLLM) {
+        String llmResponse = responseFromLLM.text();
+        logger.debugf("LLM output: %s", llmResponse);
+
+        Object result = deserialize(llmResponse);
+        if (result != null) {
+            return successWith(llmResponse, result);
+        }
+
+        String json = guardrailsUtils.trimNonJson(llmResponse);
+        if (json != null) {
+            result = deserialize(json);
+            if (result != null) {
+                return successWith(json, result);
+            }
+        }
+
+        return reprompt("Invalid JSON",
+                "Make sure you return a valid JSON object following "
+                        + "the specified format");
+    }
+
+    private Object deserialize(String llmResponse) {
+        return getOutputClass() != null ? guardrailsUtils.deserialize(llmResponse, getOutputClass())
+                : guardrailsUtils.deserialize(llmResponse, getOutputType());
+    }
+
+    protected Class<?> getOutputClass() {
+        return null;
+    }
+
+    protected TypeReference<?> getOutputType() {
```

Comment:
Wondering if we can use `Type` to merge both the output class and output type?

Reply:
I tried to, but `TypeReference` doesn't extends `Type`, but wraps it so I didn't find a reasonable way to do this.